flink/src/main/java/org/apache/iceberg/flink/IcebergTableSink.java

```diff
@@ -59,12 +60,22 @@ public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
         .map(UniqueConstraint::getColumns)
         .orElseGet(ImmutableList::of);
 
-    return (DataStreamSinkProvider) dataStream -> FlinkSink.forRowData(dataStream)
-        .tableLoader(tableLoader)
-        .tableSchema(tableSchema)
-        .equalityFieldColumns(equalityColumns)
-        .overwrite(overwrite)
-        .build();
+    return (DataStreamSinkProvider) dataStream ->  {
+      // For CDC case in FlinkSQL, change log will be rebalanced(default partition strategy) distributed to Filter opr
+      // when set job default parallelism greater than 1. That will make change log data disorder and produce a wrong
+      // result for iceberg(e.g. +U comes before -U). Here try to specific the Filter opr parallelism same as it's
+      // input to keep Filter chaining it's input and avoid rebalance.
+      Transformation<?> forwardOpr = dataStream.getTransformation();
+      if (forwardOpr.getName().equals("Filter") && forwardOpr.getInputs().size() == 1) {
+        forwardOpr.setParallelism(forwardOpr.getInputs().get(0).getParallelism());
```

Comment:
How does setting the parallelism here interplay with the parallelism of the writer?

I notice there's one other place where we set the parallelism based on some feature of the input, but we also have write parallelism that I've commented on below. Is it possible for `writeParallelism` in the below section to be different than what users have, say in the case of the CDC stream? For example, could you draw the DAG if a users input `Filter` has p=2, and then `writeParallelism` is set to `p=3`. Or is that not a possible case?

Reply:
`Filter` parallelism different than `IcebergStreamingWriter` is possible and common. Here setting `Filter` parallelism as it's input is want to let `Filter` chain to it's Input, and keep input stream data original order. `writeParallelism` will setting the `IcebergStreaming` parallelism, that will only effect on write performance. 

If `equalityFieldColumns` has been setting, that is mean data stream will be `keyBy` by `equalityFieldColumns` between `Filter` and `IcebergStreamingWriter`. If `equalityFieldColumns` is null, that will be `rebalance` between `Filter` and `IcebergStreamingWriter`.

Reply:
This is what worries me a little. this might make sense for the filter operator for the CDC stream. but is it always a good thing for other scenarios for the Flink Iceberg sink to change the parallelism of an upstream operator (which doesn't below to the Iceberg sink)?

Reply:
Resolve for https://github.com/apache/iceberg/issues/2918#issuecomment-926804817